### PR TITLE
[JENKINS-35333] New job popup

### PR DESCRIPTION
**Pull request by samatdav**
This Pull Request implements the ability to create a new job in a popup dialog window directly from home page. 
The change uses [remodal](https://github.com/VodkaBears/Remodal) library which uses css in an added file and required js from npm module.
The interface of the popup is taken from existing newJob.jelly structure and placed into a popup.
You can see the demonstration of the change in the GIF below.

popup

### CODE REVIEW

File: war/src/main/webapp/css/style.css [hunk 1]
```diff
@@ -1891,3 +1891,28 @@ body.no-sticker #bottom-sticker {
...
...6bb6a}.remodal-cancel{color:#fff;background:#e57373}.remodal-cancel:focus,.remodal-cancel:hover{background:#ef5350}.remodal-cancel::-moz-focus-inner,.remodal-close::-moz-focus-inner,.remodal-confirm::-moz-focus-inner{padding:0;border:0}@-webkit-keyframes remodal-opening-keyframes{from{-webkit-transform:scale(1.05);transform:scale(1.05);opacity:0}to{-webkit-transform:none;transform:none;opacity:1}}@keyframes remodal-opening-keyframes{from{-webkit-transform:scale(1.05);transform:scale(1.05);opacity:0}to{-webkit-transform:none;transform:none;opacity:1}}@-webkit-keyframes remodal-closing-keyframes{from{-webkit-transform:scale(1);transform:scale(1);opacity:1}to{-webkit-transform:scale(.95);transform:scale(.95);opacity:0}}@keyframes remodal-closing-keyframes{from{-webkit-transform:scale(1);transform:scale(1);opacity:1}to{-webkit-transform:scale(.95);transform:scale(.95);opacity:0}}@-webkit-keyframes remodal-overlay-opening-keyframes{from{opacity:0}to{opacity:1}}@keyframes remodal-overlay-opening-keyframes{from{opacity:0}to{opacity:1}}@-webkit-keyframes remodal-overlay-closing-keyframes{from{opacity:1}to{opacity:0}}@keyframes remodal-overlay-closing-keyframes{from{opacity:1}to{opacity:0}}
```
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: does `@import` directive works?

File: war/src/main/webapp/css/style.css [hunk 2]
```diff
@@ -1891,3 +1891,28 @@ body.no-sticker #bottom-sticker {
...
+
+.remodal {
```
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: use bem style to describe container for this modal

**Comment by oleg-nenashev** [comment]
Please create a JIRA issue with the new feature description

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 3]
```diff
@@ -43,6 +45,55 @@ THE SOFTWARE.
         <l:task href="${rootURL}/${it.viewUrl}newJob" icon="icon-new-package icon-md" it="${app}" permission="${it.itemCreatePermission}" title="${%NewJob(it.newPronoun)}"/>
       </st:include>
 
+      <l:task href="#modal" icon="icon-new-package icon-md" title="New Item Popup"/>
+      <!-- <st:include page="popup.jelly" optional="true" /> -->
+
+      <st:include optional="true">
```
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: what include (optional), but without any other args do?

Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: Apparently, nothing. I though I needed to include html between these tags. However, it seems that I dont. I will remove them.

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: You mean importing like that? <style type="text/css" media="screen">@import "${cssURL}/popup.css";</style>
And then placing popup.css in ${cssURL} folder?
It seems that ${cssURL} is not defined when I try to include into sidepanel.jelly

Comment (lanwen) [inline war/src/main/webapp/css/style.css]: is this used somewhere else?

Comment (samatdav) [inline war/src/main/webapp/css/style.css]: I found it here https://github.com/jenkinsci/iphoneview-plugin/blob/master/src/main/resources/hudson/plugins/iphoneview/tags/layout.jelly

**Comment by samatdav** [comment]
@oleg-nenashev I created a JIRA issue. https://issues.jenkins-ci.org/browse/JENKINS-35333
Please comment if there is anything wrong.

File: war/src/main/webapp/css/style.css (hunk 2 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: you mean to add new classes to use bem selectors? Or to rename existing ones? Since it is build both on remodal and bootstrap libs there are many existing css rules.

Comment (lanwen) [inline war/src/main/webapp/css/style.css]: All of this rules u should rewrite with bem. To use remodal inside of bem-block.

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 4]
```diff
@@ -43,6 +45,52 @@ THE SOFTWARE.
...
+      <l:task href="#modal" icon="icon-new-package icon-md" title="New Item Popup"/>
+
+      <div class="remodal" data-remodal-id="modal">
+        <div class="container" id="add-item-panel" style="display: none;">
+          <div class="row">
+            <div class="col-md-offset-2 col-md-20">
+              <form method="post" action="createItem" name="createItem" id="createItem">
+                <div class="header">
+                  <div class="add-item-name">
+                    <label for="name">${%ItemName.label}</label>
```
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: So where can I look to fix replacements of ${%ItemName.label} by a corresponding string?

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: So is it a right way way to import the library? Should CSS and JS be imported the same way? So is there a need to use js-modules? Tom talked about creation of detached version of remodal library - to not use jQuery in a global scope. Should I do it?

File: core/src/main/resources/hudson/model/View/popup.jelly [hunk 5]
```diff
@@ -0,0 +1,43 @@
+<div class="remodal" data-remodal-id="modal">
```
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/popup.jelly]: Are there suggestions on what can I do to include this popup.jelly file to sidepanel.jelly file? Not just inserting the remodal div into sidepanel.jelly as it is now. As discussed [here](https://github.com/samatdav/jenkins/commit/a82ebdbb31063604ad4b6e043232dc632b2068e9) with Tom, using <st:include page="popup.jelly"  /> does not quite work and throws an [error loading popup.jelly](https://gist.github.com/samatdav/4c51a0787eeda12460750520e7974245). The popup.jelly is placed in the same folder as sidepanel.jelly

File: war/src/main/webapp/css/style.css [hunk 6]
```diff
@@ -1891,3 +1891,28 @@ body.no-sticker #bottom-sticker {
...
+  max-width: 800px;
+}
+.row .col-md-offset-2 {
+  margin-left: 0;
+}
+.row .col-md-20 {
+  padding-left: 0;
+  padding-right: 0;
+  width: 100%
+}
```
Comment (recena) [inline war/src/main/webapp/css/style.css]: 👎 Please, please!!! Do not override styles by the CSS framework. This is BIG hack!

**Comment by recena** [comment]
This PR should not be merged until this https://github.com/jenkinsci/jenkins/pull/2372 is finished.

File: war/src/main/webapp/css/style.css (hunk 6 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: Sure, that was just a quick demo code, which I want to change. You suggest styling by addition of new classes? As in bem style?

**Comment by recena** [comment]
@samatdav 

1. Include the JIRA ID in the PR title
1. If you are working on [JENKINS-35333](https://issues.jenkins-ci.org/browse/JENKINS-35333), assign it yourself and update the state of the issue.

File: war/src/main/js/add-item.js [hunk 7]
```diff
@@ -107,7 +107,6 @@ $.when(getItems()).done(function(data) {
           btn.removeClass('disabled');
           btn.prop('disabled', false);
         }
-        btn.focus();
```
Comment (recena) [inline war/src/main/js/add-item.js]: Why?

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (recena) [inline war/src/main/webapp/css/style.css]: 👎 These are not global styles

File: war/src/main/webapp/css/style.css (hunk 6 above)
Comment (recena) [inline war/src/main/webapp/css/style.css]: You do not need to do that. Simply, to use the framework in a right way.

Comment (recena) [inline war/src/main/webapp/css/style.css]: `bem`?

**Comment by recena** [comment]
@samatdav FYI https://wiki.jenkins-ci.org/display/JENKINS/Instructions+for+Committers

File: war/src/main/js/add-item.js (hunk 7 above)
Comment (samatdav) [inline war/src/main/js/add-item.js]: If the GET request sent after each character and all three checks pass (js allows to create a job) then the enableSubmit function is called and OK button is focused after each character typed.
Kirill suggested not to remove this focus and create different functions for blur and keyup events. So this is not finished also.

File: war/src/main/webapp/css/style.css (hunk 6 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: As suggested above by Kirill to use bem style to use remodal inside of bem-block
Here is bem website: https://en.bem.info/methodology/key-concepts/

File: war/src/main/js/add-item.js (hunk 7 above)
Comment (samatdav) [inline war/src/main/js/add-item.js]: I included JIRA ID in PR title and assigned myselft in JIRA

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: @recena what do you suggest to do?

File: war/src/main/js/add-item.js (hunk 7 above)
Comment (recena) [inline war/src/main/js/add-item.js]: 👎 The current behavior was deeply discussed with @gusreiber. At any rate, it seems there are a lot of code in progress. I'll wait until something more stable.

File: war/src/main/webapp/css/style.css (hunk 6 above)
Comment (recena) [inline war/src/main/webapp/css/style.css]: Ah ok. I did not know about BEM, it seems something logical and reasonable.

File: war/src/main/js/add-item.js (hunk 7 above)
Comment (samatdav) [inline war/src/main/js/add-item.js]: Sure, I do not insist on my changes - just a suggestion. I think that check of name validity after each character is a useful feature. Since most of the issues with new job name check were solved recently, this is my only suggestion for the improvement left.

File: war/src/main/webapp/css/style.css (hunk 2 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: I am sorry I don't quite get how should I do it. Since I did not modify form itself, it seems that I only modify nested blocks: https://gist.github.com/samatdav/ca8085e4e117edbc25a010b6174386d1
However, are blocks named any different in bem if there are no elements and modifiers?

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: You should save library css in separate file - thats why you need import statement.

File: war/src/main/webapp/css/style.css (hunk 2 above)
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: [RU About bem](https://www.youtube.com/watch?v=LznY9VUK-h8&index=11&list=PLLkvpHo_HuBMU_fM4v-VS5VbUi9QuKyDR)

You should describe your container with help of bem methodology. Then simply use remodal inside of your container.

File: core/src/main/resources/hudson/model/View/popup.jelly (hunk 5 above)
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/popup.jelly]: try to place it in path same as class name in exception

File: war/src/main/webapp/css/style.css (hunk 6 above)
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: getbem.com is better site to describe this methodology

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: Yes, I tried to put css to another file, but could not import it (got file "404 not found").
Can I use `<style> @import url('containingFolder/popup.css'); </style>` ?
If so, what `containingFolder` would be?

File: core/src/main/resources/hudson/model/View/popup.jelly (hunk 5 above)
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/popup.jelly]: You mean where `hudson.model.AllView` class is located? I tried to put popup.jelly to folders model, model/View, model/AllView but with the same error. Header part of the form is inserted without errors, but a bigger part crashes.

File: war/src/main/webapp/css/style.css (hunk 6 above)
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: Here the source for the site https://github.com/getbem/getbem.com/

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 4 above)
Comment (tfennelly) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: What is `${%ItemName.label} ` supposed to be? The `%` looks dodgy. Rings JSP bells.

`ItemName` does not look like an instance id of any kind. It looks more like a raw type name. What are you expecting to happen here? Where are you expecting the values to come from?

Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: `${%ItemName.label} ` is localization marker. Try to search `.properties` files with such key

Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: Yes, that was it. I fixed the problem by adding newJob properties to sidepanel.properties

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 8]
```diff
@@ -43,6 +46,47 @@ THE SOFTWARE.
         <l:task href="${rootURL}/${it.viewUrl}newJob" icon="icon-new-package icon-md" it="${app}" permission="${it.itemCreatePermission}" title="${%NewJob(it.newPronoun)}"/>
       </st:include>
 
+      <l:task href="#modal" icon="icon-new-package icon-md" title="Remodal Popup"/>
+      <!-- <st:include page="popup.jelly" optional="true" /> -->
+      
+      <div class="remodal new-job-popup" data-remodal-id="modal" id="add-item-panel">
```
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: @lanwen is this better now? I successfully removed bootstrap components (container, row, etc.) and made the popup work without it. You can also see that I changed styles. And now the only rule I need is for .new-job-popup

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (samatdav) [inline war/src/main/webapp/css/style.css]: It seems that now I have only one rule now. Should I put it in add-item.css ?  Seems like a good place for it.
However, I still don't quite understand how to import remodal library rules. Are there any existing examples of importing css like that?

Comment (tfennelly) [inline war/src/main/webapp/css/style.css]: I'd think you'll need to bundle the remodal CSS inside in Jenkins and use an `adjunct` or `<l:css src="..." />`

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 8 above)
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: what for you need `remodal` class?

File: war/src/main/webapp/css/style.css (hunk 1 above)
Comment (lanwen) [inline war/src/main/webapp/css/style.css]: Maybe as example:
https://github.com/jenkinsci/github-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/github/admin/GitHubHookRegisterProblemMonitor/index.groovy#L13

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 8 above)
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: I think this is how the library references element. Without this class the popup does not even open. I believe this is like bootstrap modal which also contains .modal in it.

Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: can you point lib to search `.new-job-popup`?

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 9]
```diff
@@ -33,6 +33,10 @@ THE SOFTWARE.
     <link rel="alternate" title="Jenkins:${it.viewName} (all builds) (RSS 2.0)" href="${rootURL}/${it.url}rssAll?flavor=rss20" type="application/rss+xml" />
     <link rel="alternate" title="Jenkins:${it.viewName} (failed builds)" href="${rootURL}/${it.url}rssFailed" type="application/rss+xml" />
     <link rel="alternate" title="Jenkins:${it.viewName} (failed builds) (RSS 2.0)" href="${rootURL}/${it.url}rssFailed?flavor=rss20" type="application/rss+xml" />
+
+    <l:css src="jsbundles/remodal.css" />
```
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: I removed new styles from style.css and added remodal library styles to remodal.css
Is this correct?

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 10]
```diff
@@ -33,6 +33,10 @@ THE SOFTWARE.
...
+    <l:js src="jsbundles/add-item.js" />
+    <l:css src="jsbundles/add-item.css" />
```
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: I tried to add .new-job-popup style to add-item.css
However, every time I rebuild jenkins, all my changes from my file are removed. Could you explain why? It is strange to me since that was not the case with add-item.js

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 8 above)
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: I probably don't know how to do it yet. Do you have ideas how?
I always used library styles by referring to their class and though this is how they should be used.

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 11]
```diff
@@ -43,6 +47,47 @@ THE SOFTWARE.
         <l:task href="${rootURL}/${it.viewUrl}newJob" icon="icon-new-package icon-md" it="${app}" permission="${it.itemCreatePermission}" title="${%NewJob(it.newPronoun)}"/>
       </st:include>
 
+      <l:task href="#modal" icon="icon-new-package icon-md" title="Remodal Popup"/>
+      <!-- <st:include page="popup.jelly" optional="true" /> -->
```
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: So how can I include this popup.jelly ?
I tried to put popup.jelly to folders model, model/View, model/AllView but the error is the same. Header part of the form is inserted without errors, but a bigger part crashes.

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 10 above)
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: where exactly you put your code? (full path)
Maybe it combined from different sources as bundle with help of gulp?

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 8 above)
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: What documentation of lib says about retargeting? Did you tried to read the library code?

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 10 above)
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: oh, I found it. It is bacause add-item.css is compiled from add-item.less
This less file imports addform.less where I added my rule.

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 9 above)
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: think it not correct. Place it to some vendor dir close to less, and try to find how to process this file to jsbundles

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 11 above)
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: maybe you should make your popup.jelly as regular jelly file, but not only body part (add <jelly... to the begining)

Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: Great, thanks! That worked indeed. I will commit the change now.

**Comment by lanwen** [comment]
Can you detach commits about new look and just make this pr as popup? Just to be prepared to merge. (it should pass tests at minimal state). Then new look should be in separate pr on top of this pr, as of it can take some long time to be reviewed by core-maintainers.

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 12]
```diff
@@ -43,6 +47,9 @@ THE SOFTWARE.
         <l:task href="${rootURL}/${it.viewUrl}newJob" icon="icon-new-package icon-md" it="${app}" permission="${it.itemCreatePermission}" title="${%NewJob(it.newPronoun)}"/>
       </st:include>
 
+      <l:task href="#modal" icon="icon-new-package icon-md" title="Remodal Popup"/>
```
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: just replace existing - no need to duplicate

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 13]
```diff
@@ -42,12 +42,7 @@ THE SOFTWARE.
     <l:tasks>
       <st:include page="tasks-top.jelly" it="${it.owner}" optional="true" />
 
-      <st:include page="tasks-new.jelly" it="${it.owner}" optional="true">
-        <!-- made overridable for ViewGroup that doesn't allow modifications -->
```
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: Why you throw away this logic?

Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: oh, my bad. I missed it. I updated the file so the only change is the `href="#modal"`

**Comment by samatdav** [comment]
I created two additional pull requests (at samatdav/jenkins) to have different PRs for different features.
* This PR is for popup modal dialog
* [New Job Name Validation](https://github.com/samatdav/jenkins/pull/3/files) PR is for improvements in javascript client side name validation
* [New Job UI Change](https://github.com/samatdav/jenkins/pull/2/files) PR is for new proposed style changes

**Comment by samatdav** [comment]
@lanwen It seems that many tests fail because of Remodal JavaScript library inclusion. Where and how I should include it? The popup works locally by using two lines:
```
//jQuery min js
//Remodal min js
```

**Comment by lanwen** [comment]
It would be nice to attach remodal.js to commit if you want to bundle it to jenkins. Now don't see it in any place

**Comment by lanwen** [comment]
Also you should put your `war/src/main/webapp/jsbundles/remodal.css` to smth like `war/src/main/js/widgets/add/remodal.css` and be sure it copied to `jsbundle` location. 
As of `jsbundle` should not contain code - it should be completely autofilled place.

**Comment by lanwen** [comment]
Please provide current view in description (you have gif recorded)

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 9 above)
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: I moved the file to src/main/js/remodal.css and added changes to gulpfile.js to process it to jsbundles. I asked Tom if I placed the file correctly and if changes in gulpfile are valid. He approved it.

**Comment by samatdav** [comment]
@lanwen please take a look at the new changes. With the help of Tom, the remodal js and css files are used properly. Now, all [the tests](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/5536/) are passed without failures.
However, just as in [name validation](https://github.com/jenkinsci/jenkins/pull/2416) the build still fails. Why is that?
What would you suggest to do next?

File: war/hs_err_pid3655.log [hunk 14]
```diff
@@ -0,0 +1,692 @@
+#
```
Comment (lanwen) [inline war/hs_err_pid3655.log]: why you've commited this file?

Comment (michaelneale) [inline war/hs_err_pid3655.log]: .gitignore needs to exclude .log files or hs*log at least

File: core/src/main/resources/hudson/model/View/sidepanel.jelly [hunk 15]
```diff
@@ -33,16 +33,22 @@ THE SOFTWARE.
...
+    <l:css src="jsbundles/add-item.css" />
   </l:header>
   <l:side-panel>
     <l:tasks>
       <st:include page="tasks-top.jelly" it="${it.owner}" optional="true" />
 
       <st:include page="tasks-new.jelly" it="${it.owner}" optional="true">
         <!-- made overridable for ViewGroup that doesn't allow modifications -->
-        <l:task href="${rootURL}/${it.viewUrl}newJob" icon="icon-new-package icon-md" it="${app}" permission="${it.itemCreatePermission}" title="${%NewJob(it.newPronoun)}"/>
+        <l:task href="#modal" icon="icon-new-package icon-md" it="${app}" permission="${it.itemCreatePermission}" title="${%NewJob(it.newPronoun)}"/>
```
Comment (lanwen) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: can you name it "#newjob"?

**Comment by lanwen** [comment]
You have a problem with the build, please fix it

```
09:43:21 [INFO] [09:43:21] LESS CSS pre-processing completed to 'src/main/webapp/jsbundles'.
09:43:22 [INFO] src/main/js/add-item.js: line 3, col 5, '$remodal' is defined but never used.
09:43:22 [INFO] 
09:43:22 [INFO] 1 error
09:43:24 [ERROR] 
09:43:24 [ERROR] events.js:141
09:43:24 [ERROR]       throw er; // Unhandled 'error' event
09:43:24 [ERROR]       ^
09:43:24 [ERROR] Error: JSHint failed for: src/main/js/add-item.js
09:43:24 [INFO] ------------------------------------------------------------------------
09:43:24 [INFO] Reactor Summary:
09:43:24 [INFO] 
09:43:24 [INFO] Jenkins main module ................................ SUCCESS [  6.981 s]
09:43:24 [INFO] Jenkins cli ........................................ SUCCESS [ 12.323 s]
09:43:24 [INFO] Jenkins core ....................................... SUCCESS [03:55 min]
09:43:24 [INFO] Jenkins war ........................................ FAILURE [01:20 min]
09:43:24 [INFO] Tests for Jenkins core ............................. SKIPPED
09:43:24 [INFO] ------------------------------------------------------------------------
09:43:24 [INFO] BUILD FAILURE
09:43:24 [INFO] ------------------------------------------------------------------------
09:43:24 [INFO] Total time: 05:38 min
09:43:24 [INFO] Finished at: 2016-06-29T09:43:24-04:00
09:43:24 [INFO] Final Memory: 53M/747M
09:43:24 [INFO] ------------------------------------------------------------------------
09:43:24 [ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.0:gulp (gulp bundle) on project jenkins-war: Failed to run task: 'gulp.js jshint bundle' failed. (error code 1) -> [Help 1]
```

**Comment by lanwen** [comment]
Also i can't see remodal-detached here: https://github.com/jenkinsci/js-libs

File: war/hs_err_pid3655.log (hunk 14 above)
Comment (samatdav) [inline war/hs_err_pid3655.log]: That was an accident. I did not change gitignore settings so I don't know why that was commited. Now I removed it. I think may be I will make a new pull requst later with one commit.

**Comment by samatdav** [comment]
@tfennelly should I make a pull request to https://github.com/jenkinsci/js-libs add remodal-detached there?

**Comment by samatdav** [comment]
@lanwen do you get this build failure with `mvn jenkins-dev:run` ? This how the error looked until the last fix of remodal-detached. Now I do not get this error. Can you tell how to reproduce it?

**Comment by lanwen** [comment]
@samatdav this error in the PR build log
image

**Comment by samatdav** [comment]
@lanwen can you suggest how to debug it? Since this error does not appear locally anymore.
It is strange that is is the exact same error as in https://github.com/jenkinsci/jenkins/pull/2416
However, the changes in failing add-item.js are very different. 
I am interested if the build still fail if I remove all changes from add-item.js

File: core/src/main/resources/hudson/model/View/sidepanel.jelly (hunk 15 above)
Comment (samatdav) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly]: ok, I did.

**Comment by lanwen** [comment]
> @tfennelly should I make a pull request to https://github.com/jenkinsci/js-libs add remodal-detached there?

don't ask, just send

> how to debug it?

add logs, try to think why it can be so

**Comment by samatdav** [comment]
@lanwen could you suggest what could lead to this error? `throw er; // Unhandled 'error' event`
Googling did not really help. Some answers [say](http://stackoverflow.com/questions/34290796/events-js-141-throw-er-unhandled-error-event) that other applications may run on the same port. But as far as I know, that is not relevant in my case.

**Comment by samatdav** [comment]
I am also a bit confused with another test failing.
```
notAllowedName(hudson.model.ViewTest)
java.lang.AssertionError: ".." should not be allowed.
```
As described in [ViewTest.java](https://www.codatlas.com/github.com/jenkinsci/jenkins/HEAD/test/src/test/java/hudson/model/ViewTest.java), the test sets job name to ".." and checks if the creation is blocked.
It is strange that the test fails here since
1. I did not do changes in add-item.js - client side validation works unchanged
2. I did not do changes in server side checks.

**Comment by michaelneale** [comment]
@samatdav the error in the PR validation is a jshint failure:
```
04:16:57 [INFO] src/main/js/add-item.js: line 3, col 5, '$remodal' is defined but never used.
04:16:57 [INFO] 
04:16:57 [INFO] 1 error
04:16:59 [ERROR] 
04:16:59 [ERROR] events.js:141
04:16:59 [ERROR]       throw er; // Unhandled 'error' event
04:16:59 [ERROR]       ^
04:16:59 [ERROR] Error: JSHint failed for: src/main/js/add-item.js
```

npm isn't being so helpful there, but I am guessing it is the $remodal not being used that is actually the error.

**Comment by lanwen** [comment]
Some of the tests may fail if selectors has been changed, please check what failed test do

**Comment by samatdav** [comment]
@tfennelly could you hint how to resolve this JSHint failure? It seems like this exactly the error you helped me to solve locally, changing the remodal npm module. 
It is strange that it says `'$remodal' is defined but never used.`

File: war/src/main/js/add-item.js [hunk 16]
```diff
@@ -1,6 +1,6 @@
 // Initialize all modules by requiring them. Also makes sure they get bundled (see gulpfile.js).
 var $ = require('jquery-detached').getJQuery();
-var $remodal = require('remodal-detached').getRemodal();
+require('remodal-detached').getRemodal();
```
Comment (michaelneale) [inline war/src/main/js/add-item.js]: why is this required but then not used?

Comment (samatdav) [inline war/src/main/js/add-item.js]: This is like a bootstrap library. I do not need to use the rules by myself. But I need predifined rules for the standard behaviour

Comment (michaelneale) [inline war/src/main/js/add-item.js]: ok 👍🏻

**Comment by samatdav** [comment]
It may be ready for merge at this point.

**Comment by daniel-beck** [comment]
Triggering a new PR build.

Has anyone tested whether this retains the behavior of adding the newly created job to the current (non-"All") view?

**Comment by michaelneale** [comment]
lint error `07:38:37 [INFO] src/main/js/add-item.js: line 3, col 5, '$remodal' is defined but never used.`

**Comment by michaelneale** [comment]
ping @samatdav

**Comment by michaelneale** [comment]
as mentioned by @daniel-beck this needs to honour current behavior of adding things to the currently selected view

**Comment by lanwen** [comment]
@samatdav did you fix problem related to default view when its not "All"?

**Comment by samatdav** [comment]
@oleg-nenashev @lanwen @recena @michaelneale @tfennelly @recena Could you please review the pull request? 
@recena do you have objections against remodal library? If so what do you suggest to do?

**Comment by lanwen** [comment]
LGTM

**Comment by michaelneale** [comment]
Current state of this is that remodal should not be used (cc @recena).

**Comment by oleg-nenashev** [comment]
IIRC the decision was not to merge it in the current state

**Comment by daniel-beck** [comment]
This PR appears to have been abandoned.
